clients/wallet/src/stellar_wallet.rs

```diff
@@ -374,12 +378,92 @@ impl StellarWallet {
 			operations,
 		)?;
 
-		let _ = self.cache.save_tx_envelope(envelope.clone())?;
-
 		self.submit_transaction(envelope).await
 	}
 }
 
+// Error handling methods for StellarWallet.
+impl StellarWallet {
+	pub async fn handle_error(&self, error: Error) -> Result<TransactionResponse, Error> {
+		match &error {
+			Error::HorizonSubmissionError { title: _, status: _, reason, envelope_xdr } =>
+				if let Some(env_xdr) = envelope_xdr {
+					let transaction_envelope =
+						TransactionEnvelope::from_base64_xdr(env_xdr).expect("should convert ok");
+					if !self.is_valid_seq_number(&transaction_envelope).await {
+						if let Some(seq_number) = transaction_envelope.sequence_number() {
+							let _ = self.cache.remove_tx_envelope(seq_number);
+						}
+						return Err(error)
+					}
+					match &reason[..] {
+						"tx_bad_seq" =>
+							self.bump_sequence_number_and_submit(transaction_envelope).await,
+						"tx_internal_error" => self.submit_transaction(transaction_envelope).await,
+						_ => Err(error),
+					}
+				} else {
+					Err(error)
+				},
+
+			Error::CacheError(CacheError {
+				kind: CacheErrorKind::SequenceNumberAlreadyUsed,
+				path: _,
+				envelope,
+				sequence_number: _,
+			}) => {
+				if let Some(transaction_envelope) = envelope {
+					return self.bump_sequence_number_and_submit(transaction_envelope.clone()).await
+				}
+				Err(error)
+			},
+			_ => Err(error),
+		}
+	}
+
+	fn get_transaction_from_envelope(
+		&self,
+		tx_envelope: TransactionEnvelope,
+	) -> Option<Transaction> {
+		match tx_envelope {
+			TransactionEnvelope::EnvelopeTypeTxV0(transaction) => Some(transaction.tx.into()),
+			TransactionEnvelope::EnvelopeTypeTx(transaction) => Some(transaction.tx),
+			_ => None,
+		}
+	}
+
+	async fn is_valid_seq_number(&self, tx_envelope: &TransactionEnvelope) -> bool {
+		let account = self
+			.client
+			.get_account(self.get_public_key(), self.is_public_network)
+			.await
+			.expect("Expected account");
+		if let Some(sequence_number) = tx_envelope.sequence_number() {
+			return sequence_number > account.sequence
+		}
+		false
+	}
+
+	async fn update_transaction_sequence_number(&self, tx: Transaction) -> Transaction {
```

Comment:
You're right, @b-yap, I agree. I used this name because perhaps we could implement a different update logic, one that doesn't involve simply adding 1 to the sequence number. But then again, maybe we won't.